pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,202 @@
+	ip, ipnet, err := ipAllocator.parseCIDR(cidr)
+	if err != nil {
+		return "", err
+	}
+	var reservedIPRanges = make(map[string]bool)
+
+	// The final reserved list is obtained by combining the cloudInstancesReservedIPRanges list and the pendingIPRanges list in the ipAllocator
+	for cloudInstancesReservedIPRange := range cloudInstancesReservedIPRanges {
+		reservedIPRanges[cloudInstancesReservedIPRange] = true
+	}
+
+	// Lock is placed here so that the pendingIPRanges list captures all the IPs pending reservation in the cloud instances
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	for reservedIPRange := range ipAllocator.pendingIPRanges {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	for cidrIP := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIP) && err == nil; cidrIP, err = incrementIP(cidrIP, incrementStep29IPRange) {
+		overLap := false
+		for reservedIPRange := range reservedIPRanges {
+			_, reservedIPNet, err := net.ParseCIDR(reservedIPRange)
+			if err != nil {
+				return "", err
+			}
+			// Creating IPnet object using IP and mask
+			cidrIPNet := &net.IPNet{
+				IP:   cidrIP,
+				Mask: ipRangeMask,
+			}
+
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap, err = isOverlap(cidrIPNet, reservedIPNet)
+
+			// Error while processing ipnet
+			if err != nil {
+				return "", err
+			}
+			if overLap {
+				break
+			}
+		}
+		if !overLap {
+			ipRange := fmt.Sprint(cidrIP.String(), "/", ipRangeSize)
+			ipAllocator.holdIPRange(ipRange)
+			return ipRange, nil
+		}
+	}
+
+	// No unreserved IP range available in the entire CIDR range since we did not return
+	return "", fmt.Errorf("all of the /29 IP ranges in the cidr %s are reserved", cidr)
+}
+
+// isOverlap checks if two ipnets have any overlapping IPs
+func isOverlap(ipnet1 *net.IPNet, ipnet2 *net.IPNet) (bool, error) {
+	if ipnet1 == nil || ipnet2 == nil {
+		return true, fmt.Errorf("invalid ipnet object provided for cidr overlap check")
+	}
+	return ipnet1.Contains(ipnet2.IP) || ipnet2.Contains(ipnet1.IP), nil
+}
+
+// ParseCIDR function parses the CIDR and returns the ip and ipnet object if the cidr is valid
+// For a CIDR to be valid it must satisfy the following properties
+// 1) Network address bits must be less than 30
+// 2) The IP in the CIDR must be 'aligned' i.e we must have 8 available IPs before byte overflow occurs
+func (ipAllocator *IPAllocator) parseCIDR(cidr string) (net.IP, *net.IPNet, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, nil, err
+	}
+	// The reserved-ipv4-cidr network size must be at least /29
+	cidrSize, _ := ipnet.Mask.Size()
+	if cidrSize > ipRangeSize {
+		return nil, nil, fmt.Errorf("the maximum size of network address in the cidr can be /%d", ipRangeSize)
```

Comment:
"maximum" is sort of misleading because the bigger the number, the smaller the network is. Maybe instead the message should say "the reserved-ipv4-cidr network size must be at least /29"